File: 1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res;
    int freq[10];
    
    void solve(TreeNode* root) {
        if (root == nullptr) {
            return;
        }
        freq[root->val]++;
        if (root->left == nullptr && root->right == nullptr) {
            int odd = 0;
            for (auto it : freq) {
                if (it & 1) {
                    odd++;
                }
            }
            if (odd <= 1) {
                res++;
            }
        }
        solve(root->left);
        solve(root->right);
        freq[root->val]--;
    }
    
    int pseudoPalindromicPaths (TreeNode* root) {
        res = 0;
        solve(root);
        return res;
    }
};

File: 1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int result;
    public int[] freq;
    
    public int pseudoPalindromicPaths(TreeNode root) {
        this.result = 0;
        this.freq = new int[10];
        solve(root);
        return result;
    }
    
    public void solve(TreeNode root) {
        if (root == null) {
            return;
        }
        freq[root.val] += 1;
        if (root.left == null && root.right == null) {
            this.result += hasAtmostOneOddFreqVal(freq) ? 1 : 0;
        }
        solve(root.left);
        solve(root.right);
        freq[root.val] -= 1;
    }
    
    public boolean hasAtmostOneOddFreqVal(int[] freq) {
        int count = 0;
        for (int i = 0; i < 10; i++) {
            if (freq[i] % 2 == 1) {
                count += 1;
            }
        }
        return count <= 1;
    }
}